### ISSUE TITLE
Memory not freed without explicit drop

### ISSUE THREAD

**Issue by bhansconnect** [tracel-ai/burn#4302]
**Describe the bug**
I am still not sure exactly why this happens. In a long running program, this presents as a memory leak. That said, if the program is short enough, the memory will cleanup at the end of the program. Maybe with the full backend or device clean?

Essentially, using `.clone().select(...)` on a larger tensor to create minibatches in a loop will lead to leaking memory on each iteration unless the larger tensors are explicitly dropped (despite rust theoretically dropping them either way). At least this seems related to the root cause, I had to add it for the issue to properly reproduce. These are InnerBackend tensors, so they shouldn't have autodiff or gradients, but creating them this way and then running autodiff training seems to produce the issues.

I believe this is related to autodiff. In [my full repo](https://github.com/bhansconnect/burn-ppo), the leak happens during the training phase and tracy reports each allocation that leaks as coming from the autodiff backend. And some leaks came from optimizer.step. I am still working on a larger workaround that functions for my entire repo, but so far not having luck in the full setup.

It really feels like this accumulation should not happen. Or at least at some point it should get clean up to avoid consuming gigabytes of data in a long running program. If this is expected, it should somehow be documented better to help others avoid this class of bug.

**To Reproduce**
Use these two files as `Cargo.toml` and `src/main.rs`. Simply `cargo run --release`. It will print out memory stats. It shows about 4MB of extra data accumulating each iterations. You can uncomment the explicit drops and the accumulation goes away.

Cargo.toml:
```toml
[package]
name = "burn-mem-leak"
version = "0.1.0"
edition = "2021"
description = "Minimal reproducer for memory leak in Burn autodiff"

[dependencies]
burn = { version = "0.19", features = ["autodiff", "ndarray"] }
stats_alloc = "0.1"
rand = "0.8"
```

src/main.rs:
```rust
//! Minimal reproducer for memory accumulation in Burn autodiff
//!
//! This demonstrates how the `clone().select()` pattern causes memory to
//! accumulate during PPO-style training loops until source tensors are dropped.
//!
//! THE ISSUE:
//! - Source tensors (obs_inner, etc.) are created once and live across all epochs/minibatches
//! - Each minibatch uses `tensor.clone().select(0, indices)` to extract a subset
//! - These cloned selections keep references to the original tensor's data
//! - Memory accumulates during the loops (~4 MB per update in our tests)
//! - Memory is only freed when the source tensors are dropped (at end of function/scope)
//!
//! THE FIX:
//! - Explicitly `drop()` source tensors after the loops
//! - Then call `sync()` and `memory_cleanup()` to release memory immediately
//!
//! Run with: cargo run --release

use burn::backend::{Autodiff, NdArray};
use burn::module::Module;
use burn::nn::{Linear, LinearConfig};
use burn::optim::{AdamConfig, GradientsParams, Optimizer};
use burn::prelude::*;
use burn::tensor::activation::relu;
use burn::tensor::backend::Backend as BackendTrait;
use rand::seq::SliceRandom;
use rand::Rng;
use stats_alloc::{Region, StatsAlloc, INSTRUMENTED_SYSTEM};
use std::alloc::System;

#[global_allocator]
static GLOBAL: &StatsAlloc<System> = &INSTRUMENTED_SYSTEM;

type InnerBackend = NdArray;
type Backend = Autodiff<InnerBackend>;

const OBS_DIM: usize = 128;
const HIDDEN_DIM: usize = 512;
const ACTION_DIM: usize = 64;
const TOTAL_BATCH: usize = 8192;
const NUM_EPOCHS: usize = 4;
const NUM_MINIBATCHES: usize = 4;
const MINIBATCH_SIZE: usize = TOTAL_BATCH / NUM_MINIBATCHES;
const NUM_UPDATES: usize = 10;
const CLIP_EPSILON: f32 = 0.2;

#[derive(Module, Debug)]
struct SimpleMLP<B: burn::tensor::backend::Backend> {
    fc1: Linear<B>,
    fc2: Linear<B>,
    fc3: Linear<B>,
}

impl<B: burn::tensor::backend::Backend> SimpleMLP<B> {
    fn new(device: &B::Device) -> Self {
        Self {
            fc1: LinearConfig::new(OBS_DIM, HIDDEN_DIM).init(device),
            fc2: LinearConfig::new(HIDDEN_DIM, HIDDEN_DIM).init(device),
            fc3: LinearConfig::new(HIDDEN_DIM, ACTION_DIM).init(device),
        }
    }

    fn forward(&self, x: Tensor<B, 2>) -> Tensor<B, 2> {
        let x = relu(self.fc1.forward(x));
        let x = relu(self.fc2.forward(x));
        self.fc3.forward(x)
    }
}

fn compute_loss<B: burn::tensor::backend::AutodiffBackend>(
    model: &SimpleMLP<B>,
    obs: Tensor<B::InnerBackend, 2>,
    actions: Tensor<B::InnerBackend, 1, Int>,
    old_log_probs: Tensor<B::InnerBackend, 1>,
    advantages: Tensor<B::InnerBackend, 1>,
) -> Tensor<B, 1> {
    let logits = model.forward(Tensor::from_inner(obs));
    let log_probs = burn::tensor::activation::log_softmax(logits, 1);
    let indices_2d = actions.unsqueeze_dim(1);
    let action_log_probs: Tensor<B, 1> = log_probs
        .gather(1, Tensor::from_inner(indices_2d))
        .squeeze_dims(&[1]);

    let log_ratio = action_log_probs - Tensor::from_inner(old_log_probs).detach();
    let ratio = log_ratio.exp();

    let neg_adv: Tensor<B, 1> = Tensor::from_inner(-advantages).detach();
    let loss1 = neg_adv.clone() * ratio.clone();
    let loss2 = neg_adv * ratio.clamp(1.0 - CLIP_EPSILON, 1.0 + CLIP_EPSILON);

    loss1.max_pair(loss2).mean()
}

fn random_obs(
    device: &<InnerBackend as BackendTrait>::Device,
    rng: &mut impl Rng,
) -> Tensor<InnerBackend, 2> {
    let data: Vec<f32> = (0..TOTAL_BATCH * OBS_DIM)
        .map(|_| rng.gen_range(-1.0..1.0))
        .collect();
    Tensor::<InnerBackend, 1>::from_floats(data.as_slice(), device).reshape([TOTAL_BATCH, OBS_DIM])
}

fn random_1d(
    device: &<InnerBackend as BackendTrait>::Device,
    rng: &mut impl Rng,
) -> Tensor<InnerBackend, 1> {
    let data: Vec<f32> = (0..TOTAL_BATCH).map(|_| rng.gen_range(-1.0..1.0)).collect();
    Tensor::<InnerBackend, 1>::from_floats(data.as_slice(), device)
}

fn random_actions(
    device: &<InnerBackend as BackendTrait>::Device,
    rng: &mut impl Rng,
) -> Tensor<InnerBackend, 1, Int> {
    let data: Vec<i32> = (0..TOTAL_BATCH)
        .map(|_| rng.gen_range(0..ACTION_DIM as i32))
        .collect();
    Tensor::from_ints(data.as_slice(), device)
}

#[allow(clippy::cast_possible_wrap)]
fn main() {
    println!("Burn Memory Accumulation Reproducer");
    println!("====================================");
    println!(
        "Batch: {}, Epochs: {}, Minibatches: {}",
        TOTAL_BATCH, NUM_EPOCHS, NUM_MINIBATCHES
    );
    println!();

    let global_region = Region::new(&GLOBAL);
    let device = Default::default();
    let mut model: SimpleMLP<Backend> = SimpleMLP::new(&device);
    let mut optimizer = AdamConfig::new().init();
    let mut rng = rand::thread_rng();

    println!("{:>5} {:>12} {:>12}", "Upd", "Net (MB)", "Total (MB)");
    println!("{}", "-".repeat(31));

    let mut cumulative_net: isize = 0;

    for update in 0..NUM_UPDATES {
        let update_region = Region::new(&GLOBAL);

        // =======================================================================
        // SOURCE TENSORS: Created here, live until end of this loop iteration.
        // The clone().select() pattern below keeps references to these tensors,
        // causing memory to accumulate during the epoch/minibatch loops.
        // =======================================================================
        let obs_inner = random_obs(&device, &mut rng);
        let actions_inner = random_actions(&device, &mut rng);
        let old_log_probs_inner = random_1d(&device, &mut rng);
        let advantages_inner = random_1d(&device, &mut rng);

        for _epoch in 0..NUM_EPOCHS {
            let mut indices: Vec<usize> = (0..TOTAL_BATCH).collect();
            indices.shuffle(&mut rng);

            for mb_idx in 0..NUM_MINIBATCHES {
                let mb_start = mb_idx * MINIBATCH_SIZE;
                let mb_end = mb_start + MINIBATCH_SIZE;

                let mb_indices: Vec<i64> = indices[mb_start..mb_end]
                    .iter()
                    .map(|&i| i as i64)
                    .collect();
                let mb_indices_inner: Tensor<InnerBackend, 1, Int> =
                    Tensor::from_ints(mb_indices.as_slice(), &device);

                // ===============================================================
                // THE PROBLEMATIC PATTERN: clone().select()
                // Each call creates allocations that reference the source tensor.
                // Memory accumulates here and won't be freed until source drops.
                // ===============================================================
                let mb_obs = obs_inner.clone().select(0, mb_indices_inner.clone());
                let mb_actions = actions_inner.clone().select(0, mb_indices_inner.clone());
                let mb_old_log_probs = old_log_probs_inner
                    .clone()
                    .select(0, mb_indices_inner.clone());
                let mb_advantages = advantages_inner.clone().select(0, mb_indices_inner);

                let loss =
                    compute_loss(&model, mb_obs, mb_actions, mb_old_log_probs, mb_advantages);
                let grads = loss.backward();
                let grads_params = GradientsParams::from_grads(grads, &model);
                model = optimizer.step(1e-3, model, grads_params);
            }
        }

        // =======================================================================
        // THE FIX: Explicitly drop source tensors, then cleanup.
        // Uncomment these lines to see memory stay flat instead of accumulating.
        // Theoretically, these drop at the end of the loop any. Really not sure
        // why explicit drop is required to avoid accumulating for the whole
        // program life.
        // =======================================================================
        // drop(obs_inner);
        // drop(actions_inner);
        // drop(old_log_probs_inner);
        // drop(advantages_inner);

        let stats = update_region.change();
        let net = stats.bytes_allocated as isize - stats.bytes_deallocated as isize;
        cumulative_net += net;

        println!(
            "{:>5} {:>+12.2} {:>12.2}",
            update + 1,
            net as f64 / 1024.0 / 1024.0,
            cumulative_net as f64 / 1024.0 / 1024.0
        );
    }

    println!("{}", "-".repeat(31));
    let final_stats = global_region.change();
    let retained = final_stats.bytes_allocated as isize - final_stats.bytes_deallocated as isize;
    println!(
        "\nFinal: {:.2} MB retained ({} allocs not freed)",
        retained as f64 / 1024.0 / 1024.0,
        final_stats.allocations as isize - final_stats.deallocations as isize
    );
}
```

**Expected behavior**
The memory should get freed each iterations when the tensors are dropped. Maybe when it drops the default way, some sort of ARC issues or Autodiff gradient keeps it alive. So only explicit drops actually free the memory?

**Desktop (please complete the following information):**
 - OS: M1 MacOS with burn 0.19.1

**Comment by bhansconnect** [comment]
Oh, also, program outputs.

Without explicit drops:
```
Burn Memory Accumulation Reproducer
====================================
Batch: 8192, Epochs: 4, Minibatches: 4

  Upd     Net (MB)   Total (MB)
-------------------------------
    1        +8.49         8.49
    2        +4.28        12.78
    3        +4.27        17.05
    4        +4.30        21.35
    5        +4.26        25.61
    6        +4.27        29.88
    7        +4.33        34.21
    8        +4.26        38.48
    9        +4.27        42.74
   10        +4.26        47.01
-------------------------------

Final: 5.76 MB retained (2949 allocs not freed)
```

With explicit drops:
```
Burn Memory Accumulation Reproducer
====================================
Batch: 8192, Epochs: 4, Minibatches: 4

  Upd     Net (MB)   Total (MB)
-------------------------------
    1        +4.37         4.37
    2        +0.16         4.53
    3        +0.14         4.67
    4        +0.17         4.84
    5        +0.14         4.98
    6        +0.15         5.13
    7        +0.21         5.33
    8        +0.14         5.47
    9        +0.14         5.61
   10        +0.14         5.75
-------------------------------

Final: 5.75 MB retained (2919 allocs not freed)
```

If the issue is fixed, it should work like the explicit drop without requiring the explicit dorps.

**Comment by bhansconnect** [comment]
Oh, note, I think leaks happen on multiple backends, but this only catches them with the ndarray backend.
